proposals/0000-local-modules.rst

```diff
@@ -0,0 +1,1038 @@
+  the module being imported. If this lookup fails, then report an error.
+  
+* **Imports**: An ``import`` statement names a module *M* and (optionally) an 
+  import-specifier *impspec*.
+
+  Let *X* be the export environment of *M*.
+  
+  Let *I* be the interpretation of *impspec* with respect to *X*. If
+  *impspec* is omitted, let *I = X*.
+  
+  Let the *module alias A* be the module name appearing after the ``as`` in
+  the ``import`` statement. If there is no ``as``, then the module alias *A*
+  is just *M*.
+
+  The ``import`` statement brings names into scope by adding entries
+  to the global environment. Concretely, add *A . I* to the environment.
+  If the ``import`` statement does not use
+  ``qualified``, additionally add *I* to the global environment.
+
+Core Proposal Change Specification
+----------------------------------
+
+This proposal's specification is divided into pieces. This core piece is a necessary
+component of the overall proposal. Later pieces can be chosen piecemeal.
+
+1. **Definitions**
+
+   A. A qualified name can now have any number of module qualifications (instead
+      of just 0 or 1), each
+      separated by a ``.`` and with no intervening whitespace. Formally, a qualified
+      name is the pair of a list (perhaps empty) of module names and an occurrence name.
+
+      The concrete can lead
+      to potential ambiguity: is the name ``Data.List.length`` qualified by
+      ``Data.List`` (a well-formed module name) or qualified both by ``Data``
+      and by ``List``? If ``length`` is in scope qualified both by ``Data.List``
+      and by both ``Data`` and ``List``, the occurrence ``Data.List.length`` is
+      an error. There is no way to disambiguate locally without using module
+      aliases or other renamings. (Programmers should seek to avoid this scenario.)
+
+   #. An original name is now a pair of the name of the module that
+      defines it and a qualified name.
+
+      For instance in
+
+      ::
+
+         module Data.Foo where
+
+            module Bar where
+              module Baz where
+                x = 1
+
+      The original name of the ``x`` definition is ``(Data.Foo, Bar.Baz.x)``
+
+   #. The exports of a module are now a full environment: a map from qualified name to
+      original names. This is a strict broadening of the current definition of
+      an export environment.
+
+   #. An *import specifier*, or *impspec*, is now a set of qualified names, instead
+      of just a set of occurrence names. (Recall that a qualified name can have 0
+      qualifications, so this change is a broadening of the definition of an *impspec*
+      and is backward-compatible.)
+
+   #. Generalize the ``.`` and ``∂`` operations to both accept and return
+      environments. This is possible now that an export environment is just
+      a regular environment::
+
+         modid . E = { modid.x ↦ orig | (x ↦ orig) ∈ E }
+        
+         ∂(modid) E = { x ↦ orig | (modid.x ↦ orig) ∈ E }
```

Comment:
Taking into account my suggestion above that what we care about is the monoid structure of module names, this could be directly defined in your _Background_ section.